Fix namespacing for multiple instances of gazebo_ros2_control plugin

bobbleballs opened the pull request:
Fixes the issue described in #127 
It appears that when adding `__ns` tags to ros args on nodes that have already been namespaced, something odd happens.
I'm not too sure what happens on the rclcpp node side of things but removing the additional namespace tags on the args seems to fix the issue for me at least.
I'm running Humble on 22.04 so I've only tested against the humble branch.
I also removed the additional RCL_ROS_ARGS_FLAG in the arguments because I don't think it was doing anything.
Cheers Ben

tonynajjar commented:
Hi @bobbleballs, usually PRs are opened targeting master and are then backported to older branches

bobbleballs commented:
Hi @tonynajjar sorry, I had some issues with rebasing. I think I've sorted it now!

ahcorde reviewed (changes requested):
can you provide a simple example of how to use the namespace with your changes? Maybe add some text in the README.md

bobbleballs commented:
Hi @ahcorde,
I haven't changed any of the ways that you would use Namespaces before with this plugin, but now it works with multiple instances of the same plugin/robot description.
I've added to the [README.md here](https://github.com/bobbleballs/gazebo_ros2_control/tree/namespace_issue#multiple-namespaces)
Thanks!

splatter96 commented:
@ahcorde Could you please take another look at this?
I was stuck on this issue for multiple days until i stumbled upon the previously mentioned [thread](https://github.com/ros-controls/gazebo_ros2_control/pull/181). The patch in this MR fixed my problem with multi robot simulations in gazebo. It would be nice if this patch can be incorporated into the upcoming releases.

DoppiaEffe94 commented:
Hi All!
Has the mentioned fork already been merged into the master branch?
I'm having issue as well on creating multiple instances of the same robot in gazebo simulation.
Thanks.

christophfroehlich commented:
@mauriz @splatter96 @DoppiaEffe94 can you please add your review to this PR (In the "Files changed" tab in the upper right corner: "Review Changes"? Compile the plugin from source, and test it with different setups.

christophfroehlich reviewed (approved):
The description sound reasonable to me.
I tested the existing examples (without any namespaces), maybe anyone can provide a working example with two namespaced robots?

mauriz commented on doc/index.rst:
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
+
+Multiple Namespaces
+-----------------------------------------------------------
+The gazebo_ros2_control plugin can be launched in multiple namespaces. This is useful if it is desired to have multiple robots of the same description running at the same time.
```
The gazebo_ros2-control plugin supports multiple namespaces

mauriz replied:
Remove, the previous sentence is enough

mauriz commented on doc/index.rst:
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
+Multiple Namespaces
+-----------------------------------------------------------
+The gazebo_ros2_control plugin can be launched in multiple namespaces. This is useful if it is desired to have multiple robots of the same description running at the same time.
+If the namespace is to be defined in the URDF file, then you can use the ``<namespace>`` tag as detailed below:
```
It is possible to directly define the namespace in the urdf file, as showed below:

mauriz commented on doc/index.rst:
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
+    </plugin>
+  </gazebo>
+
+Where ``r1`` is the namespace that the plugin nodes will be run in. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
```
associated to the particular instance of the plugin

mauriz commented on doc/index.rst:
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
+
+Where ``r1`` is the namespace that the plugin nodes will be run in. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
+
+A second method for launching in namespaces is by using the Gazebo ``spawn_entity.py`` tool and setting the ``-robot_namespace`` parameter to the script:
```
A second method to launch an instance with a particular namespace is..

mauriz commented on doc/index.rst:
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
+
+.. code-block:: bash
+
+  ros2 run gazebo_ros spawn_entity.py -entity robot_1 -robot_namespace r1 -topic robot_description
```
use r1 as above

mauriz reviewed (commented):
Sorry that is a quick review, just minor comments. See if that is reasonable

christophfroehlich commented:
> Sorry that is a quick review, just minor comments. See if that is reasonable

Thanks for proof-reading. There is a little button in the code comment UI to include the code inside your comment, where you can propose your changes. This is then a one-click action to apply your changes for us, could you please change your comments above? 
Have you actually tested the new version, does it fix the mentioned problem?

mauriz commented:
Ok sorry I am new to this review system. I will do that as well as code checking asap, no earlier than a couple of days though.. sorry

DoppiaEffe94 commented:
Hi @christophfroehlich !

I had the time to test the compilation and running of this PR.

Somehow the provided fork from does not compile in my ROS2 Humble version. 
For sake of precision, I just forked the package from the "humble" [branch](https://github.com/ros-controls/gazebo_ros2_control/tree/humble) and removed the mentioned [lines](https://github.com/ros-controls/gazebo_ros2_control/pull/181/commits/a18aeb6b61aaf865f2208628153dd66081796e3a), without touching the README.md for now.

With the above steps, now the package compiles without any issue and the namespacing problem seems to be solved, as you can see below:

Example for husky robot with namespace "husky_1":
![gazebo_ros2_control_husky_1](https://github.com/ros-controls/gazebo_ros2_control/assets/105363148/02f2d7b5-5f30-4758-a0c3-0c2672e4a114)

and "husky_2":
![gazebo_ros2_control_husky_2](https://github.com/ros-controls/gazebo_ros2_control/assets/105363148/978d8ef2-ddb5-4911-b2da-aa6edb5faf16)

You can see [here](https://github.com/DoppiaEffe94/gazebo_ros2_control/tree/namespace_issue_humble) my fork.

Do you want me to open a new PR for the merge request?

Thank you

bobbleballs commented:
> Hi @christophfroehlich !
> 
> I had the time to test the compilation and running of this PR.
> 
> Somehow the provided fork from does not compile in my ROS2 Humble version. For sake of precision, I just forked the package from the "humble" [branch](https://github.com/ros-controls/gazebo_ros2_control/tree/humble) and removed the mentioned [lines](https://github.com/ros-controls/gazebo_ros2_control/pull/181/commits/a18aeb6b61aaf865f2208628153dd66081796e3a), without touching the README.md for now.
> 


Hi @DoppiaEffe94,
Could you give details on why this branch doesn't compile? It's currently merging against master so that could be why. This will be fixed when we backport.

Unfortunately I've now moved on from the position that I was at that was allowing me to test this regularly. But it was working when I left! I don't currently have a ROS install or the time to check things.

If possible we should test against master or rolling, but I understand that can be difficult.
Hopefully we can get this merged soon as it's a blocker for quite a few workflows!

machineIntelligence commented:
Hi there @christophfroehlich. I am using Humble and Gazebo 11 and I am trying to get a gazebo_ros2_control example running using the Plankton-uuv_simulator repo found [here](https://github.com/davidorchansky/Plankton). I can run all of the examples in the repo, but none of them use ROS2 control. Here is a snippet of my URDF 
`    <xacro:zray_base
      namespace="$(arg namespace)"
      inertial_reference_frame="$(arg inertial_reference_frame)">
      <gazebo>
        <plugin name="$(arg namespace)_uuv_plugin" filename="libuuv_underwater_object_ros_plugin.so">
          <fluid_density>1028.0</fluid_density>
          <namespace>namespace</namespace>
          <flow_velocity_topic>hydrodynamics/current_velocity</flow_velocity_topic>
          <debug>0</debug>
          <!-- Adding the hydrodynamic and hydrostatic parameters for the vehicle-->
          <xacro:zray_hydro_model namespace="$(arg namespace)"/>
        </plugin>
      
        <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
          <parameters>$(find zray_description)/config/controller_configuration.yaml</parameters>
            <robot_param>$(arg namespace)/robot_description</robot_param>
            <robot_param_node>$(arg namespace)/robot_state_publisher</robot_param_node>   
          <ros>
            <namespace>$(arg namespace)</namespace>
            <remapping>/tf:=tf</remapping>
            <remapping>/tf_static:=tf_static</remapping>
          </ros>             
        </plugin>
     </gazebo>
  </xacro:zray_base>`

After launching, if I check the node list I see /zray/robot_state_publisher, but I am getting the error 

> [gzserver-2] [ERROR] [1710189413.862622929] [zray.gazebo_ros2_control]: zray/robot_state_publisher service not available

 Additionally, checking ros2 service list I see 

> /zray/robot_state_publisher/describe_parameters
/zray/robot_state_publisher/get_parameter_types
/zray/robot_state_publisher/get_parameters
/zray/robot_state_publisher/list_parameters
/zray/robot_state_publisher/set_parameters
/zray/robot_state_publisher/set_parameters_atomically

christophfroehlich commented:
@machineIntelligence you haven't written if you have compiled the code from this PR or speaking of the released version? From your gzserver output, it seems that there is a leading `/` missing: `zray/robot_state_publisher` instead of `/zray/robot_state_publisher`.

christophfroehlich commented:
@DoppiaEffe94 how far did you get with it? could you provide a demo for gz_ros2_control_demos including two robot descriptions already provided with this package?

DoppiaEffe94 commented:
Hello @christophfroehlich.

Sorry for the delay, I was really busy and with a lot of projects going on.

I will set a demo in the next days and get back to you.

ros-gy commented:
Leaving a note here, because this is a fix that is needed for my multi-robot simulation on Humble and 22.04. @ahcorde @DoppiaEffe94 Any progress lately?

christophfroehlich commented on doc/index.rst:
```diff
@@ -281,6 +281,38 @@ The following is a basic configuration of the controllers:
+Multiple Namespaces
+-----------------------------------------------------------
+The gazebo_ros2_control plugin can be launched in multiple namespaces. This is useful if it is desired to have multiple robots of the same description running at the same time.
+If the namespace is to be defined in the URDF file, then you can use the ``<namespace>`` tag as detailed below:
```
```suggestion
It is possible to directly define the namespace in the URDF file, as shown below:
```

christophfroehlich commented on doc/index.rst:
```diff
@@ -281,6 +281,38 @@ The following is a basic configuration of the controllers:
+    </plugin>
+  </gazebo>
+
+Where ``r1`` is the namespace that the plugin nodes will be run in. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
```
```suggestion
Where ``r1`` is the namespace associated to the particular instance of the plugin. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
```

christophfroehlich commented on doc/index.rst:
```diff
@@ -281,6 +281,38 @@ The following is a basic configuration of the controllers:
+
+Where ``r1`` is the namespace that the plugin nodes will be run in. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
+
+A second method for launching in namespaces is by using the Gazebo ``spawn_entity.py`` tool and setting the ``-robot_namespace`` parameter to the script:
```
```suggestion
A second method to launch an instance with a particular namespace is by using the Gazebo ``spawn_entity.py`` tool and setting the ``-robot_namespace`` parameter to the script:
```

christophfroehlich reviewed (approved):
Tested it once again with the two documented approaches, and pushed my examples as well.

christophfroehlich commented:
@mergifyio backport humble
